src/connector/src/sink/big_query.rs

```diff
@@ -82,23 +86,103 @@ pub struct BigQueryCommon {
     pub dataset: String,
     #[serde(rename = "bigquery.table")]
     pub table: String,
-    #[serde(rename = "bigquery.max_batch_rows", default = "default_max_batch_rows")]
-    #[serde_as(as = "DisplayFromStr")]
-    pub max_batch_rows: usize,
-    #[serde(rename = "bigquery.retry_times", default = "default_retry_times")]
-    #[serde_as(as = "DisplayFromStr")]
-    pub retry_times: usize,
     #[serde(default)] // default false
     #[serde_as(as = "DisplayFromStr")]
     pub auto_create: bool,
 }
 
-fn default_max_batch_rows() -> usize {
-    1024
+struct BigQueryFutureManager {
+    // `offset_queue` holds the Some corresponding to each future.
+    // When we receive a barrier we add a Some(barrier).
+    // When we receive a chunk, if the chunk is larger than `MAX_ROW_SIZE`, we split it, and then we add n None and a Some(chunk) to the queue.
+    offset_queue: VecDeque<Option<TruncateOffset>>,
+    // When we pop a Some(barrier) from the queue, we don't have to wait `resp_stream`, we just truncate.
+    // When we pop a Some(chunk) from the queue, we have to wait `resp_stream`.
+    // When we pop a None from the queue, we have to wait `resp_stream`, but we don't need to truncate.
+    resp_stream: Pin<Box<dyn Stream<Item = Result<()>> + Send>>,
 }
+impl BigQueryFutureManager {
+    pub fn new(
+        max_future_num: usize,
+        resp_stream: impl Stream<Item = Result<()>> + Send + 'static,
+    ) -> Self {
+        let offset_queue = VecDeque::with_capacity(max_future_num);
+        Self {
+            offset_queue,
+            resp_stream: Box::pin(resp_stream),
+        }
+    }
+
+    pub fn add_offset(&mut self, offset: TruncateOffset, mut resp_num: usize) {
+        while resp_num > 1 {
+            self.offset_queue.push_back(None);
+            resp_num -= 1;
+        }
+        self.offset_queue.push_back(Some(offset));
+    }
 
-fn default_retry_times() -> usize {
-    5
+    pub async fn next_offset(&mut self) -> Result<Option<TruncateOffset>> {
+        if let Some(Some(TruncateOffset::Barrier { .. })) = self.offset_queue.front() {
+            return Ok(self.offset_queue.pop_front().unwrap());
+        }
+        self.resp_stream
```

Comment:
Here the correctness depends on that the number of non-barrier items (either None or Some(TruncateOffset::Chunk)) in the queue is the same as the inflight request. However, in the implementation, it's possible to have 0 `resp_num` when we write a chunk. For this chunk, there is no inflight request, but it will have an item in the queue, which causes inconsistency between the number of queue items and inflight requests.

I think instead of using none to represent that a chunk is split into multiple requests, we'd better store (TruncateOffset, remaining_resp_num) as the queue item. Code will be like the following
```
if let Some((offset, remaining_resp_num)) = self.offset_queue.front_mut() {
    if *remaining_resp_num == 0 {
        return Ok(self.offset_queue.pop_front().unwrap().0);
    }
    while *remaining_resp_num > 0 {
        self.resp_stream.next().await...??;
        *remaining_resp_num -= 1;
    }
} else {
    return pending().await;
}
```